src/main/java/org/myrobotlab/service/Mqtt.java

```diff
@@ -1,439 +1,538 @@
+		// This allows applications to perform asynchronous
+		// delivery without blocking until delivery completes.
+		//
+		// This sample demonstrates asynchronous deliver, registering
+		// a callback to be notified on each call to publish.
+		//
+		// The deliveryComplete method will also be called if
+		// the callback is set on the client
+		//
+		// note that token.getTopics() returns an array so we convert to a
+		// string
+		// before printing it on the console
+		log.info("Delivery complete callback: Publish Completed " + Arrays.toString(token.getTopics()));
+	}
+
+	public void disconnect() {
+		try {
+			isConnected = false;
+			MqttAsyncClient tclient = client;
+			client = null;
+			tclient.disconnect();
+			tclient.close();
+			tclient = null;
+		} catch (Exception e) {
+			// don't care
+		}
+	}
+
+	public String getTopic() {
+		return topic;
+	}
+
+	public Set<String> getSubscriptions() {
+		return subscriptions;
+	}
+
+	public String getUrl() {
+		return url;
+	}
+
+	/**
+	 * @see MqttCallback#messageArrived(String, MqttMessage)
+	 */
+	@Override
+	public void messageArrived(String topic, MqttMessage message) throws MqttException {
+		// Called when a message arrives from the server that matches any
+		// subscription made by the client
+		String time = new Timestamp(System.currentTimeMillis()).toString();
+		// FIXME - new String won't be happy with true binary payloads ..
+		String payload = new String(message.getPayload());
+		String messageStr = "onMqttMsg Time: " + time + "\tTopic: " + topic + "\tMessage: " + payload + "\tQoS: "
+				+ message.getQos();
+		log.info(messageStr);
+
+		// serialization needs to be a logical layer - so we can change it to
+		// protobuf or native Java
+		try {
+			Message msg = (Message) codec.decode(payload, Message.class);
+
+			// COMMON GATEWAY REGISTERATION AND X-FORWARDED BEGIN --------------
+
+			// make a mrl key and protocol uri begin ------
+			// ADD TCP CLIENT BEGIN ! - probably should not be in RemoteAdapter - as this is
+			// a detail for tcp
+
+			// String clientKey = String.format("mqtt://%s:%d", this.url);
+			URI uri = new URI(url);
+			// HELP PROTOKEY VS MRL KEY ??
+			// TcpThread tcp = new TcpThread(myService, uri, clientSocket);
+			// tcpClientList.put(uri, tcp);
+			// myService.connections.put(uri, tcp.data);
+
+			// make a mrl key and protocol uri end -------
+
+			// registration request - security is applied here
+			// x-forwarded is
+			// msg.name == runtime;
+
+			// what if you don't want to do x-forwarding ???
+			inbox.add(msg);
+
+			// COMMON GATEWAY REGISTERATION AND X-FORWARDED END ----------------
+
+		} catch (Exception e) {
+			error("tried decoding [%] into a message", payload);
+		}
+
+		// are all necessary ?
+		invoke("publishMqttMsg", topic, message.getPayload());
+		invoke("publishMqttMsgByte", message.getPayload());
+		invoke("publishMqttMsgString", new String(message.getPayload()), topic);
```

Comment:
Is there a reason for this?:
The first publish is `topic`, `payload`.
The third one is `payload as string`, `topic`.

Reply:
are you saying the parameters should be consistent ?  I would agree (although that wasn't part of the pr change, and it wasn't part of the scope of the ticket - but I would agree)
Or are you asking why so many different types ?